gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2

```diff
@@ -316,26 +322,56 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
                 client_cert_source = mtls.default_client_cert_source()
         return client_cert_source
 
-    def _get_api_endpoint(api_override, client_cert_source, use_mtls_endpoint):
+    def _get_api_endpoint(api_override, client_cert_source, universe_domain, use_mtls_endpoint):
         """Return the API endpoint used by the client.
 
         Args:
             api_override (str): The API endpoint override. If specified, this is always the return value of this function.
             client_cert_source (bytes): The client certificate source used by the client.
+            universe_domain (str): The universe domain used by the client.
             use_mtls_endpoint (str): How to use the MTLS endpoint, which depends also on the other parameters.
                 Possible values are "always", "auto", or "never".
 
         Returns:
             str: The API endpoint to be used by the client.
         """
-
+        use_mtls_endpoint = use_mtls_endpoint == "always" or (use_mtls_endpoint == "auto" and client_cert_source)
         if api_override is not None:
             api_endpoint = api_override
-        elif use_mtls_endpoint == "always" or (use_mtls_endpoint == "auto" and client_cert_source):
+        elif use_mtls_endpoint:
+            if universe_domain != {{ service.client_name }}.GOOGLE_DEFAULT_UNIVERSE:
+                raise MutualTLSChannelError("MTLS is not supported in any universe other than googleapis.com.")
             api_endpoint = {{ service.client_name }}.DEFAULT_MTLS_ENDPOINT
         else:
-            api_endpoint = {{ service.client_name }}.DEFAULT_ENDPOINT
+            api_endpoint = {{ service.client_name }}.DEFAULT_ENDPOINT_TEMPLATE.format(UNIVERSE_DOMAIN=universe_domain)
         return api_endpoint
+    
+    def _get_universe_domain(client_universe_domain, universe_domain_env):
+        """Return the universe domain used by the client.
+
+        Args:
+            client_universe_domain (str): The universe domain configured via the client options.
+            universe_domain_env (str): The universe domain configured via the "GOOGLE_CLOUD_UNIVERSE_DOMAIN" environment variable.
+
+        Returns:
+            str: The universe domain to be used by the client.
+        """
+        universe_domain = {{ service.client_name }}.GOOGLE_DEFAULT_UNIVERSE
+        if client_universe_domain is not None:
+            universe_domain = client_universe_domain
+        elif universe_domain_env is not None:
+            universe_domain = universe_domain_env
```

Comment:
This is a much concise alternative but I run into some issues when I make this change:
- `""` is treated as None in the `or` statement and the statement will end up evaluating to the default GDU value. As a result, the below exception for checking the empty string will never be raised (the exception is part of the design.)
- If I were to split the `or` statement to two, the `len` check on the None type would lead to an error.
- If I do the error checking before assigning the `universe_domain` value for each individual parameter, then I end up raising an exception for the following scenario i.e. `client_universe_domain="foo.com"` and `universe_domain_env=""`. In this case, `client_universe_domain` is a valid universe and we shouldn't bother if a universe is configured (even incorrectly) via the environment variable.

There are obviously workarounds to make this work but that seem to defeat the purpose of what's proposed here. 

Feel free to suggest more on this!

Reply:
Oh, good point.

I think this should work, though it's less clear than the non-working version

```suggestion
EMPTY_TOKEN=-9999 # since empty strings are falsy, but we want to distinguish them from None
universe_domain = (((client_universe_domain is not None) and (client_universe_domain or EMPTY_TOKEN)) or
                   ((universe_domain_env is not None) and (universe_domain_env or EMPTY_TOKEN)) or
                   {{ service.client_name }}.GOOGLE_DEFAULT_UNIVERSE)
 if universe_domain == EMPTY_TOKEN or len(universe_domain.strip()) == 0:
     raise ValueError(...)
```

Reply:
Your call on which of the two alternatives is easier to follow. I slightly prefer my version, but there's a tradeoff between conciseness and thinking through the code---not a big burden either way in this case.

Reply:
I think let's keep it as is since it looks easier to follow through for someone who might not be familiar with what's happening here.